internal/version/check_update.go

```diff
@@ -26,31 +45,102 @@ func CheckUpdate() {
 		return
 	}
 
-	githubClient := github.UnauthorizedClient()
-	release, _, err := githubClient.Repositories.GetLatestRelease(context.TODO(), repositoryOwner, repositoryName)
-	if err != nil {
-		logger.Debugf("Error: can't check latest release, %v", err)
+	v, ok := os.LookupEnv(checkUpdatedDisabledEnv)
+	if ok && strings.ToLower(v) != "false" {
+		logger.Debug("Disabled checking updates")
 		return
 	}
 
-	if release.TagName == nil || *release.TagName == "" {
-		logger.Debugf("Error: release tag is empty")
+	duration, err := time.ParseDuration(defaultCacheDuration)
+	if err != nil {
+		logger.Warnf("failed to parse cache duration: %s", err.Error())
 		return
 	}
+	latestVersion, expired, _ := checkCachedLatestVersion(duration)
+	var release *versionLatest
+	switch {
+	case !expired:
+		logger.Debugf("latest version (cached): %s", latestVersion)
+		release = latestVersion
+	default:
+		logger.Debugf("checking latest release in Github")
+		githubClient := github.UnauthorizedClient()
+		githubRelease, err := githubClient.LatestRelease(context.TODO(), repositoryOwner, repositoryName)
+		if err != nil {
+			logger.Debugf("Error: %v", err)
+			return
+		}
+		release = &versionLatest{
+			TagName:   *githubRelease.TagName,
+			HtmlURL:   *githubRelease.HTMLURL,
+			Timestamp: time.Now(),
+		}
+	}
 
 	currentVersion, err := semver.NewVersion(Tag[1:]) // strip "v" prefix
 	if err != nil {
 		logger.Debugf("Error: can't parse current version tag, %v", err)
 		return
 	}
 
-	releaseVersion, err := semver.NewVersion((*release.TagName)[1:]) // strip "v" prefix
+	releaseVersion, err := semver.NewVersion(release.TagName[1:]) // strip "v" prefix
 	if err != nil {
 		logger.Debugf("Error: can't parse current version tag, %v", err)
 		return
 	}
 
 	if currentVersion.LessThan(releaseVersion) {
-		logger.Infof("New version is available - %s. Download from: %s", *release.TagName, *release.HTMLURL)
+		logger.Infof("New version is available - %s. Download from: %s", release.TagName, release.HtmlURL)
+	}
+
+	if err := writeLatestReleaseToCache(release); err != nil {
```

Comment:
The file is always written even if the content is not expired?

Reply:
That's true, I've checked that scenario and it will return earlier, thanks!